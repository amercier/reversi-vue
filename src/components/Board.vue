<template>
  <div class="board">
    <div class="board-grid">
      <div
        v-for="{ row, col, isStone, isMove, color } in cells"
        :key="`${row}-${col}`"
        :class="
          `board-cell ${(row + col) % 2 === 1 ? 'board-cell--alternate' : ''}`
        "
      >
        <Stone class="board-stone" v-if="isStone" :color="color" />
        <Move
          v-if="isMove"
          class="board-move"
          :color="color"
          :onClick="() => onMove(row, col)"
        />
      </div>
    </div>
  </div>
</template>

<script>
import Stone from './Stone.vue';
import Move from './Move.vue';

/**
 * Reversi board component.
 */
export default {
  name: 'Board',
  components: {
    Stone,
    Move,
  },
  props: {
    /**
     * Array of cells.
     *
     * @type {Cell[]}
     * @see @/store For `Cell` documentation.
     */
    cells: {
      type: Array,
      required: true,
    },

    /**
     * Function to call on player click, when it is a move,
     *
     * @type {Function}
     */
    onMove: {
      type: Function,
      required: true,
    },
  },
};
</script>

<style lang="scss">
$board-padding: 1rem;
$grid-color: #1b881b;
$cell-padding: 10%;
$cell-border-thickness: 0.1rem;

.board {
  background: #c58e57 url(../assets/wooden-board.jpg) repeat 0 0;
  background-attachment: fixed;
  padding: $board-padding;
  border-radius: $board-padding;
  box-shadow: 0 0 2rem 0.5rem #000;

  &-grid {
    display: grid;
    grid-template-columns: repeat($BOARD_SIZE, 1fr);
    grid-gap: 2 * $cell-border-thickness;

    background: $grid-color url(../assets/wooden-board.jpg) repeat 0 0;
    background-attachment: fixed;
    background-blend-mode: multiply;
    background-clip: padding-box;
    overflow: hidden; // Hide cells edge borders
    border: ($cell-border-thickness) solid rgba(0, 0, 0, 0.3);
  }

  &-cell {
    // Inset effect
    margin: -$cell-border-thickness;
    border-left: $cell-border-thickness solid rgba(255, 255, 255, 0.1);
    border-top: $cell-border-thickness solid rgba(255, 255, 255, 0.1);
    border-right: $cell-border-thickness solid rgba(0, 0, 0, 0.1);
    border-bottom: $cell-border-thickness solid rgba(0, 0, 0, 0.1);

    // Maintain 1:1 ratio
    height: 0;
    padding-top: 100%;
    position: relative;

    &--alternate {
      background-color: rgba(0, 0, 0, 0.15);
    }
  }

  &-stone,
  &-move {
    position: absolute;
  }

  &-stone {
    top: $cell-padding;
    bottom: $cell-padding;
    right: $cell-padding;
    left: $cell-padding;
  }

  &-move {
    top: 0;
    bottom: 0;
    right: 0;
    left: 0;
  }
}
</style>
