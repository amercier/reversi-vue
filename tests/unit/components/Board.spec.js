import { shallowMount } from '@vue/test-utils';
import Board from '@/components/Board.vue';

describe('Board.vue', () => {
  const propsData = {
    cells: [
      { row: 0, col: 0, isStone: true, isMove: false, color: 'white' },
      { row: 0, col: 1, isStone: true, isMove: false, color: 'black' },
      { row: 1, col: 0, isStone: false, isMove: true, color: 'black' },
      { row: 1, col: 1, isStone: false, isMove: false, color: 'blank' },
    ],
  };

  it('renders without crashing', () => {
    expect(() => shallowMount(Board, { propsData })).not.toThrow();
  });

  it('renders the board', () => {
    const board = shallowMount(Board, { propsData });
    expect(board.element).toMatchSnapshot();
  });

  it('emits a "moved" event when a move is clicked', () => {
    const board = shallowMount(Board, { propsData });
    board.find('.board-move').vm.$emit('clicked');
    expect(board.emitted('moved')).toBeArray();
    expect(board.emitted('moved')).toHaveLength(1);
    expect(board.emitted('moved')[0]).toStrictEqual([1, 0]);
  });

  describe('methods', () => {
    const { methods } = Board;

    describe('position(row, col)', () => {
      const { position } = methods;

      it('returns A1 for [0,0]', () => {
        expect(position(0, 0)).toBe('A1');
      });

      it('returns H1 for [0,7]', () => {
        expect(position(0, 7)).toBe('H1');
      });

      it('returns A8 for [7,0]', () => {
        expect(position(7, 0)).toBe('A8');
      });

      it('returns H8 for [7,7]', () => {
        expect(position(7, 7)).toBe('H8');
      });
    });
  });
});
