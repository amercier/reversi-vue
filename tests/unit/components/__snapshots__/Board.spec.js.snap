// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board.vue renders the board 1`] = `
<div
  class="board"
>
  <div
    class="board-grid"
  >
    <div
      class="board-cell "
    >
      <stone-stub
        class="board-stone"
        color="white"
      />
       
      <!---->
    </div>
    <div
      class="board-cell board-cell--alternate"
    >
      <stone-stub
        class="board-stone"
        color="black"
      />
       
      <!---->
    </div>
    <div
      class="board-cell board-cell--alternate"
    >
      <!---->
       
      <move-stub
        class="board-move"
        color="black"
        title="A2"
      />
    </div>
    <div
      class="board-cell "
    >
      <!---->
       
      <!---->
    </div>
  </div>
</div>
`;
